### PR TITLE
feat(parser): add functionality to self hosted images

### PR DESCRIPTION
**Pull request by chamod-gamage**
* Adds `images/update` route to parser which takes any scraped images found in posts and reuploads them to our S3 bucket (defined via env variables)
* Any unavailable images (e.g. post taken down) are deleted since the image can't be accessed anyways
* Adds `got` and `aws-sdk` for HTTP requests and piping images to s3

### CODE REVIEW

File: apps/parser/controllers/index.js [hunk 1]
```diff
@@ -95,3 +96,51 @@ export const updateAllPosts = async (req, res) => {
   }
   res.status(200).json(results);
 };
+
+const getOriginalFileName = (post, url) => {
+  let fileName = "";
+  if (url.includes("scontent")) {
+    fileName = `${post.id}/images/${url.split("?")[0].split("/").pop()}`;
```
Comment (LemonFace0309) [inline apps/parser/controllers/index.js:103]: nit: but might be nice to add example comments for what this does

File: apps/parser/controllers/index.js [hunk 2]
```diff
@@ -95,3 +96,51 @@ export const updateAllPosts = async (req, res) => {
...
+  } else if (url.includes("external")) {
+    fileName = `${post.id}/images/${url
+      .split("&url=")[1]
+      .split("&cfs=")[0]
+      .replace("%3A", ":")
+      .replace("%2F", "/")}`;
```
Comment (LemonFace0309) [inline apps/parser/controllers/index.js:109]: same here

File: apps/parser/controllers/index.js [hunk 3]
```diff
@@ -95,3 +96,51 @@ export const updateAllPosts = async (req, res) => {
...
+  for (let i = 0; i < posts.length; i++) {
+    for (let j = 0; j < posts[i].length; j++) {
+      if (
+        posts[i][j]._doc.images.length &&
+        posts[i][j]._doc.images[0].includes("uwcribs")
+      )
+        continue;
+      for (let k = 0; k < posts[i][j]._doc.images.length; k++) {
+        img = posts[i][j]._doc.images[k];
+        if (!img.includes("scontent") || !img.includes("external")) continue;
```
Comment (LemonFace0309) [inline apps/parser/controllers/index.js:130]: is this hard coded from the squidward image? What if we just host the squidward image ourselves in the bucket

**Review by LemonFace0309 — APPROVED** [review]
LGTM, some extra comments might be nice but great otherwise :) Feel free to backlog the comments I added because of time constraints

File: apps/parser/controllers/helpers.js [hunk 4]
```diff
@@ -136,3 +143,24 @@ export const extractImages = async (driver) => {
...
+      .upload({
+        Bucket: "uwcribs",
+        Key: filePath,
+        Body: response,
+        ContentType: fileType,
+      })
+      .promise();
+    return uploadedFile.Location;
+  } catch (e) {
+    return "";
```
[the remark this answers is not in the document]
Comment (chamod-gamage) [reply apps/parser/controllers/helpers.js:164]: I can log but in most cases, don't need to be concerned about errors. This was the only way to bypass dead images (link expired from scrape source).

File: apps/parser/controllers/index.js (hunk 3 above)
Comment (chamod-gamage) [reply apps/parser/controllers/index.js:130]: This was just to check whether the images are from FB, if they aren't the image is skipped.

**Pull request merged**